pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,622 @@ package awsplacementgroup
+		logger.Info(fmt.Sprintf("no matching AWS placement group for name %s", pg.Name))
+	default:
+		// Exactly 1 placement group exists with the this name,
+		// observe its configuration and set it on the object Status.
+		logger.Info(fmt.Sprintf("found 1 AWS placement group for name %s with id: %s", pg.Name, *placementGroups.PlacementGroups[0].GroupId))
+		setObservedConfiguration(pg, placementGroups.PlacementGroups[0])
+	}
+
+	// Set the .Status.ExpiresAt in 2 minutes from now, to keep a TTL cache
+	// of the configuration observed from AWS.
+	inTwoMinutes := metav1.NewTime(metav1.Now().Add(2 * time.Minute))
+	pg.Status.ExpiresAt = &inTwoMinutes
+
+	return nil
+}
+
+// setObservedConfiguration sets the configuration observed from the AWS placement group to
+// the ObservedConfiguration field in the Status of the object.
+func setObservedConfiguration(pg *machinev1.AWSPlacementGroup, placementGroup *ec2.PlacementGroup) {
+	pg.Status.ObservedConfiguration.GroupType = machinev1.AWSPlacementGroupType(strings.Title(aws.StringValue(placementGroup.Strategy)))
+	pg.Status.ObservedConfiguration.Partition = &machinev1.AWSPartitionPlacement{Count: int32(aws.Int64Value(placementGroup.PartitionCount))}
+
+	condition := metav1.Condition{Type: "Ready"}
+
+	var equal bool
+	if pg.Spec.ManagementSpec.Managed != nil {
+		equal = reflect.DeepEqual(pg.Status.ObservedConfiguration, *pg.Spec.ManagementSpec.Managed)
+	}
+
+	if equal {
+		condition.Status = metav1.ConditionTrue
+		condition.Reason = AWSPlacementGroupConfigurationInSyncConditionReason
+	} else {
+		condition.Status = metav1.ConditionFalse
+		condition.Reason = AWSPlacementGroupConfigurationMismatchConditionReason
+	}
+
+	meta.SetStatusCondition(&pg.Status.Conditions, condition)
+}
+
+// validateAWSPlacementGroup validates an AWSPlacementGroup configuration.
+func validateAWSPlacementGroup(pg *machinev1.AWSPlacementGroup) error {
+	// First validation should happen via webhook before the object is persisted.
+	// This is a complementary validation to fail early in case of lacking proper webhook validation.
+	switch pg.Spec.ManagementSpec.ManagementState {
+	case machinev1.ManagedManagementState, machinev1.UnmanagedManagementState:
+		// valid values
+	default:
+		return fmt.Errorf("invalid AWSPlacementGroup. spec.managementSpec.managementState must either be %s or %s",
+			machinev1.ManagedManagementState, machinev1.UnmanagedManagementState)
+	}
+
+	if pg.Spec.ManagementSpec.ManagementState == machinev1.ManagedManagementState {
+		if pg.Spec.ManagementSpec.Managed == nil {
+			return fmt.Errorf("invalid AWSPlacementGroup. spec.managementSpec.managed must not be nil when spec.managementSpec.managementState is %s",
+				machinev1.ManagedManagementState)
+		}
+
+		// A Managed placement group may be moved to Unmanaged, however an Unmanaged
+		// group may not be moved back to Managed.
+		if pg.Status.ManagementState == machinev1.UnmanagedManagementState {
```

Comment:
Ahh we already have the validation I asked for above, nice

Reply:
👍